### PR TITLE
* Handle vm create, get and delete

### PR DESCRIPTION
**Pull request by outscale-vbr**
* Handle link and unlink volume
* Handle add vm as backend
* Handle attach eip
* Handle add sg with vm sgmember

### CODE REVIEW

File: cloud/services/compute/vm.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+// ValidateVmType check that vmType is a valid vmType
+func ValidateVmType(vmType string) (string, error) {
+	switch {
+	case strings.Contains(vmType, "small") || strings.Contains(vmType, "nano") || strings.Contains(vmType, "micro") || strings.Contains(vmType, "medium") || strings.Contains(vmType, "large") || strings.Contains(vmType, "tina"):
+		return vmType, nil
+	default:
+		return vmType, errors.New("Invalid vmType")
+	}
+}
```
Comment (outscale-mdr) [inline cloud/services/compute/vm.go:68]: This checks is light to me

File: cloud/services/compute/vm.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	}
+
+	oscApiClient := s.scope.GetApi()
+	oscAuthClient := s.scope.GetAuth()
+	vmResponse, httpRes, err := oscApiClient.VmApi.CreateVms(oscAuthClient).CreateVmsRequest(vmOpt).Execute()
+	if err != nil {
+		fmt.Printf("Error with http result %s", httpRes.Status)
+		return nil, err
+	}
+	vmID := *(*vmResponse.Vms)[0].VmId
```
Comment (outscale-mdr) [inline cloud/services/compute/vm.go:113]: Check for nil pointer. Use maybe `VmResponse.HasVms()` and getters

File: cloud/services/compute/vm.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+	vms, ok := vmResponse.GetVmsOk()
+	if !ok {
+		return nil, errors.New("Can not get vm")
+	}
+	if len(*vms) == 0 {
+		return nil, nil
+	} else {
+		vm := *vms
+		return &vm[0], nil
+	}
```
Comment (outscale-mdr) [inline cloud/services/compute/vm.go:129]: This should be checked before vmId affectation

File: cloud/services/security/publicip.go [hunk 4]
```diff
@@ -108,3 +112,62 @@ func (s *Service) ValidatePublicIpIds(publicIpIds []string) ([]string, error) {
...
+		publicIp, err := s.GetPublicIp(publicIpId)
+		if err != nil {
+			return err
+		}
+		_, ok := publicIp.GetLinkPublicIpIdOk()
+		if !ok {
+			getPublicIpUnlink = true
+		}
+		time.Sleep(clockInsideLoop * time.Second)
+
```
Comment (outscale-mdr) [inline cloud/services/security/publicip.go:167]: You should return otherwise you will wait again

File: cloud/services/compute/vm.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+			return err
+		}
+		vmState, ok := vm.GetStateOk()
+		if !ok {
+			return errors.New("Can not get vm state")
+		}
+		if *vmState == state {
+			getVmState = true
+		}
+		time.Sleep(clockInsideLoop * time.Second)
```
Comment (outscale-mdr) [inline cloud/services/compute/vm.go:188]: You should return otherwise you will wait again

File: cloud/services/security/securitygroup.go [hunk 6]
```diff
@@ -42,14 +42,31 @@ func (s *Service) CreateSecurityGroup(netId string, securityGroupName string, se
 	return securityGroup, nil
 }
 
-func (s *Service) CreateSecurityGroupRule(securityGroupId string, flow string, ipProtocol string, ipRange string, fromPortRange int32, toPortRange int32) (*osc.SecurityGroup, error) {
+func (s *Service) CreateSecurityGroupRule(securityGroupId string, flow string, ipProtocol string, ipRange string, securityGroupMemberId string, fromPortRange int32, toPortRange int32) (*osc.SecurityGroup, error) {
+	var rule osc.SecurityGroupRule
+	if securityGroupMemberId != "" && ipRange == "" {
```
Comment (outscale-mdr) [inline cloud/services/security/securitygroup.go:47]: What if `securityGroupMemberId != "" && ipRange == ""`  ?

File: cloud/services/service/load_balancer.go [hunk 7]
```diff
@@ -237,3 +274,25 @@ func (s *Service) DeleteLoadBalancer(spec *infrastructurev1beta1.OscLoadBalancer
...
+	var getLoadBalancerDeregisterVm = false
+	if !getLoadBalancerDeregisterVm {
+		loadBalancer, err := s.GetLoadBalancer(spec)
+		if err != nil {
+			return err
+		}
+		loadBalancerBackendVmIds := loadBalancer.GetBackendVmIds()
+		if len(loadBalancerBackendVmIds) == 0 {
+			getLoadBalancerDeregisterVm = true
+		}
```
Comment (outscale-mdr) [inline cloud/services/service/load_balancer.go:291]: Same

**Review by outscale-mdr — CHANGES_REQUESTED** [review]

File: api/v1beta1/types.go [hunk 8]
```diff
@@ -293,6 +310,14 @@ var (
 	VmStateTerminated                   = VmState("terminated")
 	VmStateStopping                     = VmState("stopping")
 	VmStateStopped                      = VmState("stopped")
+	DefaultVmName                string = "cluster-api-vm"
+	DefaultVmSubregionName       string = "eu-west-2a"
+	DefaultVmImageId             string = "ami-bb490c7"
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply api/v1beta1/types.go:315]: Hi
It is already and will be  a custom image made with packer.

File: api/v1beta1/types.go [hunk 9]
```diff
@@ -275,13 +285,20 @@ type OscKeypair struct {
...
-	SubnetName          string   `json:"subnetName,omitempty"`
-	ResourceId          string   `json:"resourceId,omitempty"`
+	Name                string                 `json:"name,omitempty"`
+	ImageId             string                 `json:"imageId,omitempty"`
+	KeypairName         string                 `json:"keypairName,omitempty"`
+	VmType              string                 `json:"vmType,omitempty"`
+	VolumeName          string                 `json:"volumeName,omitempty"`
+	DeviceName          string                 `json:"deviceName,omitempty"`
+	SubnetName          string                 `json:"subnetName,omitempty"`
+	LoadBalancerName    string                 `json:"loadBalancerName,omitempty"`
```
[the remark this answers is not in the document]
Comment (outscale-vbr) [reply api/v1beta1/types.go:295]: It is a loadbalancer to have a control plane endpoint.

File: api/v1beta1/types.go [hunk 10]
```diff
@@ -275,13 +285,20 @@ type OscKeypair struct {
...
+	KeypairName         string                    `json:"keypairName,omitempty"`
+	VmType              string                    `json:"vmType,omitempty"`
+	VolumeName          string                    `json:"volumeName,omitempty"`
+	DeviceName          string                    `json:"deviceName,omitempty"`
+	SubnetName          string                    `json:"subnetName,omitempty"`
+	LoadBalancerName    string                    `json:"loadBalancerName,omitempty"`
+	PublicIpName        string                    `json:"publicIpName,omitempty"`
+	SubregionName       string                    `json:"subregionName,omitempty"`
+	PrivateIps          []OscPrivateIpElement     `json:"privateIps,omitempty"`
+	SecurityGroupsNames []OscSecurityGroupElement `json:"securityGroupsNames,omitempty"`
```
Comment (outscale-mdr) [inline api/v1beta1/types.go:299]: Change to `SecurityGroupNames` (also in json)

File: cloud/services/compute/vm.go [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+		return imageId, nil
+	default:
+		return imageId, errors.New("Invalid imageId")
+	}
+}
+
+// ValidateDeviceName check that DeviceName  is a valid DeviceName
+func ValidateDeviceName(deviceName string) (string, error) {
+	switch {
+	case strings.Contains(deviceName, "/dev/xvd") && len(deviceName) == 9:
```
Comment (outscale-mdr) [inline cloud/services/compute/vm.go:53]: Maybe check that it starts with '/dev/xvd'  and last character is a number ?

File: cloud/services/compute/vm.go [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+
+	oscApiClient := s.scope.GetApi()
+	oscAuthClient := s.scope.GetAuth()
+	vmResponse, httpRes, err := oscApiClient.VmApi.CreateVms(oscAuthClient).CreateVmsRequest(vmOpt).Execute()
+	if err != nil {
+		fmt.Printf("Error with http result %s", httpRes.Status)
+		return nil, err
+	}
+	if !vmResponse.HasVms() {
+		return nil, nil
```
Comment (outscale-mdr) [inline cloud/services/compute/vm.go:115]: This is an error, no ?

File: cloud/services/compute/vm.go [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+	}
+	vms, ok := vmResponse.GetVmsOk()
+	if !ok {
```
Comment (outscale-mdr) [inline cloud/services/compute/vm.go:118]: This check is the same as `vm.Response.HasVms()`

File: cloud/services/compute/vm.go [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+		vm, err := s.GetVm(vmId)
+		if err != nil {
+			return err
+		}
+		vmState, ok := vm.GetStateOk()
+		if !ok {
+			return errors.New("Can not get vm state")
+		}
+		if *vmState == state {
+			return nil
```
Comment (outscale-mdr) [inline cloud/services/compute/vm.go:190]: My mistake, a `break` makes more sense

File: cloud/services/security/securitygroup.go [hunk 15]
```diff
@@ -42,14 +42,32 @@ func (s *Service) CreateSecurityGroup(netId string, securityGroupName string, se
...
+			SecurityGroupId: &securityGroupMemberId,
+		}
+		rule = osc.SecurityGroupRule{
+			SecurityGroupsMembers: &[]osc.SecurityGroupsMember{securityGroupMember},
+			IpProtocol:            &ipProtocol,
+			FromPortRange:         &fromPortRange,
+			ToPortRange:           &toPortRange,
+		}
+	} else if securityGroupMemberId != "" && ipRange != "" {
+		return nil, errors.New("Missing neither ipRange or neither securityGroupMemberId")
```
Comment (outscale-mdr) [inline cloud/services/security/securitygroup.go:58]: On the contrary, both are setted

**Review by outscale-mdr — CHANGES_REQUESTED** [review]

File: cloud/services/compute/vm.go (hunk 11 above)
Comment (outscale-vbr) [reply cloud/services/compute/vm.go:53]: last character is a number is never a number ?

Comment (outscale-mdr) [reply cloud/services/compute/vm.go:53]: letter sorry :)

File: api/v1beta1/types.go (hunk 10 above)
Comment (outscale-vbr) [reply api/v1beta1/types.go]: Ok

File: cloud/services/compute/vm.go (hunk 11 above)
Comment (outscale-vbr) [reply cloud/services/compute/vm.go]: Ok

File: cloud/services/compute/vm.go (hunk 12 above)
Comment (outscale-vbr) [reply cloud/services/compute/vm.go]: Remove

File: cloud/services/compute/vm.go (hunk 14 above)
Comment (outscale-vbr) [reply cloud/services/compute/vm.go]: Ok

**Review by outscale-mdr — APPROVED** [review]

**Comment by outscale-vbr** [comment]
close https://github.com/outscale-dev/cluster-api-provider-outscale/issues/26

**Pull request merged**